brave/brave-ios
# Fix #7013: IPFS navigation in omnibox

## Summary of Changes
Brave will detect user enter ipfs scheme url and ask (default) consent to resolve ipfs scheme url via a public gateway value (configurable in Brave settings.)


This pull request fixes #7013 #7128 #7129

## Submitter Checklist:


## Test Plan:

1. Open Brave and go to Settings/Web3
2. Observe there are two new rows under IPFS. `Resolve IPFS Resources` with `Ask` as the default value. `IPFS Public Gateway Address` with `https://dweb.link` as the default value. 
3. Click the `IPFS Public Gateway Address` to another one like `https://w3s.link` to see if it gets successfully set. Go back to the previous page the new value should be displayed. 
4. Click the `IPFS Public Gateway Address for NFT` and set it to another value other than its previous value or `IPFS Public Gateway Address` and see if it gets successfully set and not effect to the value of `IPFS Public Gateway Address`

=================================================================

1. Make sure the `Resolve IPFS Resources` is set to `Ask`
2. Try to open an IPFS scheme url like `ipfs://QmX4nfgA35MiW5APoc4P815hMcH8hAt7edi5H3wXkFm485/1688.html` in a new or existed tab
3. Observe that an IPFS interstitial page is displayed
4. Check the copies
6. Click `Learn more` to see if it redirects to the correct website
7. Click `User a Public Gateway` to see if the IPFS scheme url gets resolved(now should be a https scheme url) and loads 
8. Go to settings and check `Resolve IPFS Resources` is set as `Enabled`
9. Set `Resolve IPFS Resources` to be `Ask` again
10. Back to the tab load the IPFS scheme url again
11. Observe that the IPFS interstitial page shows up again
12. This time click `Disable`
13. An error page will show up indicate IPFS is disabled. 
14. Go to settings and check `Resolve IPFS Resources` is set as `Disabled`
15. Set `Resolve IPFS Resources` to be `Enabled`
16. Back to the tab the load the url again
17. This time the ipfs scheme url should be resolved and load

=================================================================

1. Regardless the preference of `Resolve IPFS Resources`, open an ipfs scheme url in a private mode tab
2. Observe an error page will show up indicate ipfs is not allowed in private mode.

=================================================================

1. Find a website that has a link with ipfs scheme url and click it
2. Observe a new tab will be opened and perform the ipfs navigation flow based on the ipfs preference

=================================================================

1. Copy and paste an ipfs url to other app other than Brave. For example the note app
2. Click the link and see if Brave app will open and load this ipfs url based on ipfs preference


## Screenshots:


<img src=https://user-images.githubusercontent.com/1187676/226718626-f971d030-bc7d-4543-be4a-bbfeec01f99b.png> | <img src=https://user-images.githubusercontent.com/1187676/226719467-c08109fe-edd4-4fa3-aa9a-eb899e3eff2f.png> | <img src=https://user-images.githubusercontent.com/1187676/226719629-981ba667-4830-44b0-ba69-b45077f08fee.png>
--|--|--

## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
